Adding support for different weight vector types

rofinn opened the pull request:
This PR is intended to address concerns raised in issues #53 and #249 

Requirements:

- [X] WeightVec -> AbstractWeights
- [X] Weights (aka default reliability weights)
- [X] FrequencyWeights (limited to vectors of integers)
- [x] ProbabilityWeights
- [X] Allow all weights to take a corrected argument which defaults to true.
- [X] Add an exponential function which creates a set of exponential `Weights`.
- [x] Update existing test cases to work with the appropriate types and test against both corrected and uncorrected versions of the weights
- [x]  Deprecate API changes
- [x] Test deprecation changes to ensure we didn't actively break any functionality?
- [x] Check that we haven't reduced test coverage.
- [x] Remove deprecation tests.
- [x] Update docs

nalimilan commented on src/weights.jl:
```diff
@@ -2,43 +2,112 @@
+        bias::Int
+    end
+
+    immutable ProbabilityWeights{S<:Real, T<:Real, V<:RealVector} <: AbstractWeights{S, T, V}
```
It turns out "sampling weights" is a much more common expression than "probability weights". Maybe we should use the former term?

rofinn replied:
I don't have very strong feelings either way, but sampling just seemed more ambiguous to me.

ararslan replied:
I've heard both. In my line of work I don't come across these kinds of weights much so I can't speak to their prominence overall, but I think "probability weights" is what I learned to call them in grad school.

nalimilan commented on src/weights.jl:
```diff
@@ -2,43 +2,112 @@
 """
-function WeightVec{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs))
-    return WeightVec{S, eltype(vs), V}(vs, s)
+function Weights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs); corrected::Bool=true)
```
Wouldn't it make more sense to decide whether to apply the correction when calling `std`? For most functions (like `mean`), the correction does not apply and the results will be the same whatever the kind of weights you have. It seems a bit weird to require people to decide about correction in advance.

rofinn replied:
Yeah, I guess that would fit with base better. I was initially thinking that it might make sense to precompute the value, but we only use it in a few places so it probably wouldn't cost us much.

rofinn replied:
I've made this change in my next commit, but it ended up being a lot of little changes.

nalimilan commented on src/weights.jl:
```diff
@@ -2,43 +2,112 @@
 """
-weights(vs::RealVector) = WeightVec(vs)
-weights(vs::RealArray) = WeightVec(vec(vs))
+    frequency(vs)
```
`frequency` and `exponential` are really too broad terms to be claimed for weights. We should find more specific names, like `fweights` and `expweights`. One (smallà advantage is that it would mirror Stata (with also `pweights` and `aweights`).

rofinn replied:
I tend to prefer full word function names. I'd be fine with keeping `frequency` and `exponential`, but only exporting the aliases of `pweights`, `eweights`, `fweights`. That way folks can choose to use the full word versions, but we won't be cluttering up the global namespace on them.

ararslan replied:
I don't think we should have two ways to access the same function. IMO we should choose one interface and stick to it. I'm with Milan on this one; I don't think we should use `frequency` and `exponential` for these, as those names are too general. `freqweights` and `expweights` seem reasonable enough names to me.

rofinn replied:
Looks like scipy also uses the `aweights` and `fweights` names https://docs.scipy.org/doc/numpy/reference/generated/numpy.cov.html

ararslan replied:
Actually, why not just use the type constructors rather than separate functions?

rofinn replied:
Iono, cause `FrequencyWeights` seem long...?

ararslan replied:
Yeah, I suppose so. Perhaps it's best then to follow suit with Stata and SciPy and use `aweights` and `fweights`.

nalimilan commented on src/weights.jl:
```diff
@@ -2,43 +2,112 @@
 
+# TODO: constructor for ProbabilityWeights, but I'm not familiar with how bias correction works with these
+# types of weights or if bias correction even makes sense.
+# https://en.wikipedia.org/wiki/Inverse_probability_weighting
```
See http://www.stata.com/support/faqs/statistics/weights-and-summary-statistics/ for a summary regarding analytical weights (aweights) and probability weights (pweights). Wikipedia has interesting data on frequency weights and analytical/precision weights: https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Weighted_sample_variance

I'll try to find more references.

rofinn replied:
As it stands, my understanding is that these weights are intended to provide inferential statistics about the population (vs sample), so assuming the weights are correct we wouldn't want to do any bias correction? Also, I guess analytical/precision weights = reliability weights on the wiki page?

nalimilan replied:
> As it stands, my understanding is that these weights are intended to provide inferential statistics about the population (vs sample)?

Yes, probability/sampling weights reflect the fact that some population groups were over/under-sampled, so individuals do not all represent the same number of people in the target population.

> Also, I guess analytical/precision weights = reliability weights on the wiki page?

Right. The most explicit name is "inverse-variance weights", and this is what [the link](https://www.gnu.org/software/gsl/manual/html_node/Weighted-Samples.html) given on Wikipedia mentions.

nalimilan commented on src/weights.jl:
```diff
@@ -2,43 +2,112 @@
-    immutable WeightVec{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
+    abstract AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}
+
+    immutable Weights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractWeights{S, T, V}
```
I'm not sure we still need this. Do we expect people to need kinds of weights which cannot have their own type?

rofinn replied:
Well I get the impression that the kinds of weights should be kept relatively small and folks might still want to dispatch based on the storage of those weights (e.g., NullableArray vs Vector).

nalimilan replied:
What I meant is that the `Weights` family might not be needed since all use cases should fall into a well-defined weights type.

rofinn replied:
Oh, yeah, I've renamed `Weights` to `AnalyticWeights` in the next commit to be more specific. I've also added an `aweights` method and made `weights` an alias to `aweights`, but maybe I should just deprecate the `weights` method in favour of making folks think more about what type of weights they want to use?

nalimilan replied:
Yes, let's deprecate it, as that's the best way to make people discover the new weight types and choose the appropriate one for them. If it turns out some people need other weight types, they will let us know.

ararslan commented on src/deprecates.jl:
```diff
@@ -2,16 +2,16 @@ import Base.@deprecate
+@deprecate _moment3(v::RealArray, m::Real, wv::AbstractWeights) _moment3(v, wv, m)
+@deprecate _moment4(v::RealArray, m::Real, wv::AbstractWeights) _moment4(v, wv, m)
+@deprecate _momentk(v::RealArray, k::Int, m::Real, wv::AbstractWeights) _momentk(v, k, wv, m)
+@deprecate moment(v::RealArray, k::Int, m::Real, wv::AbstractWeights) moment(v, k, wv, m)
```
If `WeightVec` is going away it should have a deprecation, and if it isn't then these deprecations should maintain the existing signatures. (They've probably been deprecated long enough that we can just remove them, but that's an aside.)

rofinn replied:
Yeah, I've deprecated `WeightVec` and I was contemplating deprecating `weights` in favour of `aweights` to be more specific.

nalimilan replied:
Ideally, the deprecation warning should mention all weight types. I'm also not sure we can call `aweights` as a replacement, since the current `cov` does not apply any correction, the deprecation should do the same: maybe we need to keep the `WeightsVec` type (deprecating) to avoid breaking packages?

rofinn replied:
That seems reasonable and would also help notify people about the other changes occurring with weights without introducing breaking changes. Is it possible to print custom deprecation messages with the `@deprecate` system or would I need to add that to the deprecated `WeightVec` and `weights` methods?

rofinn replied:
How do folks want to handle internal calls to `weights(x)` if it (and `WeightVec`) are deprecated?

1. We leave the code as is and accept that functions like `wsample`, `wmedian`, etc will throw deprecation warnings when they need to creating a `WeightVec` from an array.
2. Change the default behaviour to creating `AnalyticWeights` which may break existing behaviour for some people.

nalimilan replied:
AFAIK the kind of weights doesn't matter for `wsample` or `wmedian`, it only makes a difference for `var` and similar functions. IOW the type of weights created for the former functions is an implementation detail, we can choose any type we want (though I would choose frequency weights, as they are simpler to understand).

rofinn replied:
Yeah, I guess `wsample` and `wmedian` weren't good examples. Alright, I'll switch that over to avoid the warnings. Thanks.

ararslan commented on src/weights.jl:
```diff
@@ -2,43 +2,112 @@
 
 if VERSION < v"0.6.0-dev.2123"
-    immutable WeightVec{S<:Real, T<:Real, V<:RealVector} <: RealVector{T}
+    abstract AbstractWeights{S<:Real, T<:Real, V<:RealVector} <: RealVector{T}
```
Should be `@compat abstract type ... end`. Also why is this being conditioned on the version? Looks like that predates this PR, but I don't see why we would need to condition for these definitions.

rofinn replied:
`AbstractWeights{S<:Real, T<:Real, V<:AbstractVector{T}} <: AbstractVector{T}` isn't valid syntax cause 0.5 doesn't support triangular dispatch.

ararslan replied:
Oh dang you're right. 😞

ararslan commented on src/weights.jl:
```diff
@@ -2,43 +2,112 @@
+# Arguments
+* `n::Integer`: the desired length of the `Weights`
+* `λ::Real`: is a smoothing factor or rate paremeter between 0 .. 1.
+    As this value approaches 0 the resulting weights will be almost equal(),
```
Why the parentheses after "equal"? Also should remove "is" on the previous line, "paremeter" -> "parameter", and ".." -> "and".

ararslan commented on src/weights.jl:
```diff
@@ -2,43 +2,112 @@
+    while values closer to 1 will put higher weight on the end elements of the vector.
+"""
+function exponential(n::Integer, λ::Real=0.99)
+    @assert 0 <= λ <= 1 && n > 0
```
Typically for argument checking it's better to use `condition || throw(ArgumentError("..."))`. In this case,
```julia
n > 0 || throw(ArgumentError("cannot construct weights of length < 1"))
0 <= λ <= 1 || throw(ArgumentError("smoothing factor must be between 0 and 1"))
```
That provides more descriptive error messages for the user, plus it's better to use specifically-typed exceptions where possible.

ararslan commented on src/weights.jl:
```diff
@@ -345,10 +414,10 @@ end
 """
-wmedian(v::RealVector, w::RealVector) = median(v, weights(w))
-wmedian{W<:Real}(v::RealVector, w::WeightVec{W}) = median(v, w)
+wmedian(v::RealVector, w::RealVector) = median(v, weights(w, false))
```
If the `weights` API is changing to require a boolean parameter, the existing API should have a formal deprecation.

nalimilan commented:
I have found more references for the variance correction to apply to different types of weights.

Regarding sampling/probability weights, a good reference is `svyvar` from the R survey package. What it does (in the simple case where there is no stratification) is simply to apply the Bessel correction with `n` equal to the number of non-zero weights:
```julia
n = count(!iszero, w)
sw = sum(w)
xbar = sum(w .* x)/sw
sum(w .* (x - xbar).^2)/sw * n/(n - 1)
```

(Of course this is just a Julia illustration of the algorithm. This code should be optimized to avoid making copies.)

A good reference for analytical/precision weights is [`wtd.var`](https://www.rdocumentation.org/packages/Hmisc/versions/4.0-1/topics/wtd.stats) from the Hmisc R package. For these weights, we should use the same formula as `wtd.var` when passed `normwt=TRUE`, i.e. normalize the weights so that they sum to the vector length (or maybe to the number of non-zero weights for consistency with above). Then the basic formula is:
```julia
n = length(w) # Could be count(!iszero, w) instead
w = w * n/sum(w)
sw = sum(w) # This is now equal to n, but maybe we should support non-normalized weights?
xbar = sum(w .* x)/sw
sum(w * (x - xbar).^2)/(sw - sum(w.^2)/sw)
```

This corresponds to the [Wikipedia formula](https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Weighted_sample_variance), and indeed [it was written by the same person](https://github.com/harrelfe/Hmisc/issues/22#issuecomment-94453638). Regarding the actual implementation, we could normalize the weights when creating the vector if that's always needed.

Another interesting reference (for analytical/precision weights) is SAS:
http://support.sas.com/documentation/cdl/en/proc/61895/HTML/default/viewer.htm#a002473731.htm#a000068934

nalimilan commented on src/weights.jl:
```diff
@@ -2,43 +2,112 @@
+        bias::Real
+    end
+
+    immutable FrequencyWeights{S<:Integer, T<:Integer, V<:IntegerVector} <: AbstractWeights{S, T, V}
```
I don't think we should restrict frequency weights to integers. While this seems obvious at first, formulas generalize to non-integer weights, and other software generally accepts arbitrary values. This can be useful in some cases (e.g. sometimes you need to split observations between groups). Also this will be more practical if your variable is stored as a float for some reason.

rofinn commented:
Sorry about the code churn in these commits folks. Turns out moving the `corrected` option to the stats function required changing a lot of code.

# Summary of the changes:

* Reverts many test changes from the last commit
* Changed a bunch of test cases to use `corrected=false` for now
* This included a lot of little changes to method definition and some resulting formatting changes where necessary
* Added a few test cases for the corrected variances ( I just used the code @nalimilan posted above )
* Added bias correction for ProbabililtyWeights.
* Deprecated `WeightVec`
* Added a macro for easier creation of weight types.
* Renamed weight creation function to `fweights`, `pweights`, etc.

TODO:
* Change `fweights` to accept `Real`s vs `Integer`s
* Cleanup the test cases to use testsets, checking both the corrected and uncorrected versions.
* Remove some of the 0.5.0-dev version checks since StatsBase isn't supporting those versions anyways.
* Cleanup some of the existing nested ternaries in the files that this PR touches.
* Update the documentation to reflect these changes... assuming we mostly agree on the changes.

ararslan commented on src/weights.jl:
```diff
@@ -2,44 +2,181 @@
+end
+
+"""
+    `@weights name`
```
Backticks shouldn't be used for indented blocks like this, since Markdown code formatting is applied by virtue of being indented. So this will actually show up in the docstring and docs like 
```julia
`@weights name`
```

ararslan commented on src/weights.jl:
````diff
@@ -2,44 +2,181 @@
+    bias(w::ProbabilityWeights, [corrected])
+
+```math
+\fraction{n}{∑w × (n - 1)}
````
I think this should be `\frac{n}{(n - 1) \sum w}`

ararslan replied:
(Similarly `\frac` and LaTeX commands rather than Unicode elsewhere)

rofinn replied:
I'm fine with just using LaTeX commands, but the julia docs recommend using unicode characters.

>Use Unicode characters rather than their LaTeX escape sequence, i.e. ``α = 1`` rather than ``\\alpha = 1``.

ararslan replied:
Oh, weird. Anyway, should still be `\frac` rather than `\fraction`, though I guess with the backslash escaped.

nalimilan replied:
I don't think that remark applies to `\sum`, as it does more than just printing the `∑` character: it also uses the right size for the symbol. "charaters" in this context mostly means "letters" in that context AFAICT.

ararslan commented on test/cov.jl:
```diff
@@ -88,62 +88,62 @@ end
 if VERSION < v"0.5.0-dev+679"
-    @test cov(X, wv1)           ≈ S1w ./ sum(wv1)
-    @test cov(X, wv2; vardim=2) ≈ S2w ./ sum(wv2)
+    @test cov(X, wv1; corrected=false)           ≈ S1w ./ sum(wv1)
```
Having to change all of the tests suggests that this is a breaking change. To keep from breaking users' code, we should probably make `corrected=false` the default.

rofinn replied:
I'm fine with doing that for this PR, but it would be nice to switch it back to `corrected=true` in a later release?

ararslan replied:
It's tricky to change things like this, since we'd need some kind of deprecation to avoid suddenly and silently breaking others' code. I think @nalimilan has gone through such a process recently; perhaps he could be of help here.

nalimilan replied:
As I just noted in a comment above, I think the best deprecation path is to deprecate `WeightsVec` but keep it around so that existing code does not use the correction, and point to new weights types in the deprecation warning.

rofinn replied:
@nalimilan To clarify, you'd like `var`, `std` and `cov` to always return the uncorrected version for `WeightVec`s even if `corrected=true`?

nalimilan replied:
The `corrected` argument wouldn't be supported at all for `WeightVec`, or at least and error would be raised if `corrected=true` is passed.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,181 @@
+"""
+    AnalyticWeights(vs, [wsum])
+
+Construct a `AnalyticWeights` with weight values `vs` and sum of weights `wsum`.
```
"an `AnalyticWeights` vector".

Would be nice to describe analytic weights, and mention their other names. Same for other types.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,181 @@
+@weights AnalyticWeights
+
+"""
+    AnalyticWeights(vs, [wsum])
```
`wsum=sum(vs)`, without brackets.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,181 @@
 """
+    aweights(vs)
+
+Construct a `AnalyticWeights` type from a given array.
```
type -> vector.

Maybe add something like "See the documentation for that type for more information"?

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,181 @@
+Construct a `AnalyticWeights` type from a given array.
+"""
+aweights(vs::RealVector) = AnalyticWeights(vs)
+aweights(vs::RealArray) = AnalyticWeights(vec(vs))
```
Is this method really needed? I don't think we perform this kind of conversion automatically in general.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,181 @@
+    n > 0 || throw(ArgumentError("cannot construct weights of length < 1"))
+    0 <= λ <= 1 || throw(ArgumentError("smoothing factor must be between 0 and 1"))
+    w0 = map(i -> λ * (1 - λ)^(1 - i), 1:n)
+    return weights(w0)
```
Shouldn't this be `ExponentialWeights`? But why do we need a specific weights type: don't them enter into one of the above families? Or maybe they should just use a generic weights type (without any correction)?

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,181 @@
+function bias(w::AbstractWeights, corrected=true)
+    s = sum(w)
+    if corrected
+        return inv(s * (1 - sum(normalize(values(w), 1) .^ 2)))
```
`normalize` and `.^2` are going to create copies of the weights, which should be avoided. Formula should be adapted so that no allocation happens at all. Maybe you'll need a loop for that.

rofinn commented on src/moments.jl:
```diff
@@ -15,10 +15,12 @@ whereas it's `length(x)-1` in `Base.varm`. The impact is that this is not a
 variance of the sample.
 """
-Base.varm(v::RealArray, wv::WeightVec, m::Real) = _moment2(v, wv, m)
+function Base.varm(v::RealArray, wv::AbstractWeights, m::Real; corrected=true)
```
I'm aware that the short form support line breaks, but I think that's harder for people to parse. The `function ... end` syntax with proper indentation provides a more consistent read through. I tend to have similar views on ternaries that cover multiple lines (particularly if they're nested). I'm fine with keeping to the short form if that's the desired style for this repo, but maybe we should add a style guide to make that clearer?

nalimilan replied:
I don't really care. There's a short style guide in Julia's CONTRIBUTING.md, but I don't think it's mentioned.

rofinn commented on src/moments.jl:
```diff
@@ -30,23 +32,32 @@ whereas it's `length(x)-1` in `Base.var`. The impact is that this is not a
-Base.varm!(R::AbstractArray, A::RealArray, wv::WeightVec, M::RealArray, dim::Int) =
-    scale!(_wsum_centralize!(R, @functorize(abs2), A, values(wv), M, dim, true), inv(sum(wv)))
+function Base.varm!(R::AbstractArray, A::RealArray, wv::AbstractWeights, M::RealArray, dim::Int; corrected=true)
+    scale!(
```
Again, maybe a style guide would be helpful. The argument I've heard for spreading args (for long function calls) across multiple lines is that it minimizes the size of you diffs later on. I'm not saying that's the right approach, but given that there are multiple styles floating around and the julia style guide doesn't discuss them maybe it's worth clarifying.

ararslan replied:
Everyone has a different preferred style; it's mostly about being consistent within a particular package. That said, I'd love for the style guide in the manual to be more specific about things.

nalimilan replied:
This kind of thing is worth clarifying, but I think it's pretty clear from the existing code base of JuliaStats packages that this style is against the convention.

rofinn commented on src/moments.jl:
```diff
@@ -58,200 +69,263 @@ function var!(R::AbstractArray, A::RealArray, wv::WeightVec, dim::Int; mean=noth
         s += z * z
     end
-    s / n
+    s * bias(n, corrected)
```
I agree that `bias` isn't really descriptive of what it represents anymore. Would something like `bias_factor`, `scale_factor`, `bias_coef`, `scale_coef`, etc make more sense? The nice thing about keeping a relatively general name is that we can use the same function call `var`, `std` and `cov`.

nalimilan replied:
`var` and `std` are in some way based on covariance, so `cov` should be fine.

rofinn commented on src/moments.jl:
```diff
@@ -58,200 +69,263 @@ function var!(R::AbstractArray, A::RealArray, wv::WeightVec, dim::Int; mean=noth
     end
-    s / sum(wv)
+
+    result = s * bias(wv, corrected)
```
Agreed. Sorry, this is leftover from some debugging I was doing (removed the print statements, but forgot to reverse that).

ararslan commented on src/weights.jl:
````diff
@@ -2,44 +2,181 @@
+\fraction{1}{n - 1}
+```
+"""
+bias(n::Integer, corrected=true) = inv(n - Int(corrected))
````
Out of curiosity, why is it necessary to use `1/`? `inv` should be equivalent in this case, I would think. Is it just a matter of style?

nalimilan replied:
Yeah, just for style. `?inv` says it's for matrix inverse, and I find `1/` closer to math formulas.

rofinn commented on src/weights.jl:
```diff
@@ -2,44 +2,181 @@
+    bias(n::Integer, [corrected])
 
-Construct a `WeightVec` with weight values `vs` and sum of weights `wsum`.
+Computes the corrected (default) or uncorrected bias for any `n` observations.
```
We do this in quite a few places, so I think it makes more sense to consistently do `x * some_func(...)` rather that having a bunch of conditions in the calling functions.

nalimilan replied:
Why not, but then change the description to make it clear that it returns `1` if `corrected=false`.

rofinn replied:
Well the function should be returning `1 / sum(x)` when `corrected=false`.

nalimilan replied:
OK, but that still isn't a "bias", so it should be called differently.

JeffreySarnoff commented:
I just found WeightVec and the much cleaner reworking above.  I will use the new abstraction and its support to hold the vector of weights used with rolling, windowed stats.  Usually, these weights are normalized in some way and then used.  It would be pleasant to do (should this as yet not be).
```julia
my_weights = new_abstraction_above( my_weights_as_values_in_sequence )

my_normalized_weights = normalize(my_weights)
my_normalized_weights = normalize(my_weights, p_norm = 2)

normalize!(my_weights)
normalize!(my_weights, p_norm = 1.618)
```
without reaching inside your type to access the sum

rofinn commented on src/StatsBase.jl:
```diff
@@ -17,14 +17,22 @@ module StatsBase
+    AnalyticWeights,    # the default type for representing a analytic/precision/reliability weight vectors
+    FrequencyWeights,   # the type for representing a frequency weight vectors
+    ProbabilityWeights, # the type for representing a probability/sampling weight vectors
+    ExponentialWeights, # the type for representing exponential weights
```
Forgot to delete `ExponentialWeights` from export.

rofinn commented on src/deprecates.jl:
```diff
@@ -43,3 +37,28 @@ findat(a::AbstractArray, b::AbstractArray) = findat!(Array{Int}(size(b)), a, b)
 @deprecate df(obj::StatisticalModel) dof(obj)
 @deprecate df_residual(obj::StatisticalModel) dof_residual(obj)
+
+@weights WeightVec
```
Not sure if this `WeightVec` should go here or in weights.jl

ararslan replied:
If it's deprecated it should live in this file, just needs `@deprecate` as appropriate

rofinn replied:
I'm guessing using `depwarn` inside the `WeightVec` and `weights` methods is the correct approach here since we want to deprecate the functionality rather than the signature?

nalimilan replied:
Yes. There's also `@deprecate_binding` to deprecate `WeightVec` itself.

rofinn replied:
Doesn't `@deprecate_binding` only work if you're just changing the name of the binding?

rofinn commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+(ie: [Bessel's correction](https://en.wikipedia.org/wiki/Bessel's_correction)),
+otherwise it will return ``\\frac{1}{n}``.
+"""
+cfactor(n::Integer, corrected=false) = 1 / (n - Int(corrected))
```
`cfactor` (for correction factor) seemed like the best choice for this function now.

nalimilan replied:
I'm not a fan of this name either: "factor" really is secondary here, what matters is 1) that it's a correction, 2) that it applies to var/cov/std.

ararslan replied:
`varcorrection`, maybe `biascorrection`?

rofinn replied:
`varcorrection` seems a little long, maybe `cvar`?

ararslan replied:
That's even less descriptive than `cfactor`, IMO.

rofinn replied:
Alright, `varcorrection` it is then. I don't really care about the name since we're not exporting it.

rofinn replied:
Wait, what about `varden` (variance denominator)? I feel like the fact that it takes a `corrected` argument should cover that it's a correction.

nalimilan replied:
I like `varcorrection` better. Or maybe `bessel_correction`?

rofinn commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+``\\frac{1}{\sum w}``
+"""
+cfactor(wv::AbstractWeights, ::Type{Val{false}}) = 1 / sum(wv)
+cfactor(wv::AbstractWeights, ::Type{Val{true}}) =
```
I'd understand if folks are opposed to the `Type{Val{True}}`, but this means subtypes only need to implement the correction condition.

nalimilan replied:
`Val` is really not what you want here: since the value of `corrected` will only be known at runtime, you're forcing the call to `cfactor` to go via dispatch at runtime, while it would have been inlined if you used `::Bool`.

Again, I don't think we should provide a fallback correction for `AbstractWeight`: it's unlikely to be valid, which is worse than raising an error. Creating new weight types shouldn't be a frequent need, people can afford the cost of defining this simple method if it applies.

nalimilan replied:
Oops, I had missed that you throw an error. But still I think it's better to repeat the uncorrected equation in each method than to use `Val`.

rofinn commented on test/scalarstats.jl:
```diff
@@ -66,9 +66,9 @@ z2 = [8. 2. 3. 1.; 24. 10. -1. -1.; 20. 12. 1. -2.]
-@test zscore(a)    ≈ zscore(a, mean(a), std(a))
-@test zscore(a, 1) ≈ zscore(a, mean(a,1), std(a,1))
-@test zscore(a, 2) ≈ zscore(a, mean(a,2), std(a,2))
+@test zscore(a)    ≈ zscore(a, mean(a), std(a; corrected=false))
```
The `corrected=false` is because `mean_and std`now has `corrected=false`. In order to support all of the existing behaviour we may need to mix and match which methods have `corrected=true`.

nalimilan replied:
I don't understand: why not adapt `mean_and_std` to default to `corrected=true`? To avoid breaking existing code, we could keep `corrected=false` for a while, but print a warning when `corrected` isn't specified so that we can make the switch later.

JeffreySarnoff replied:
+1

rofinn replied:
I'm clearly missing something here. How am I suppose to check if a keyword argument is set?

rofinn replied:
NOTE: We could add that warning if we chose to make `corrected` a regular (vs keyword) argument. We'd just need to have a function which doesn't take the corrected argument, so that it can print a warning and call the function with the default setting? I initially thought making `corrected` a keyword argument was the best way to stay consistent with base, but `var`, `std` and `cov` aren't even consistent with each other in base, so it might make more sense to do whatever is best for our use case.

```
std(
std(A::AbstractArray; corrected, mean) in Base at statistics.jl:244         std(iterable; corrected, mean) in Base at statistics.jl:266
std(A::AbstractArray, region; corrected, mean) in Base at statistics.jl:263
julia> cov(
cov(x::AbstractArray{T,1} where T, corrected::Bool) in Base at statistics.jl:346
cov(x::AbstractArray{T,1} where T) in Base at statistics.jl:348
cov(X::AbstractArray{T,2} where T, vardim::Int64) in Base at statistics.jl:357
cov(X::AbstractArray{T,2} where T, vardim::Int64, corrected::Bool) in Base at statistics.jl:357
cov(X::AbstractArray{T,2} where T) in Base at statistics.jl:360
cov(x::AbstractArray{T,1} where T, y::AbstractArray{T,1} where T, corrected::Bool) in Base at statistics.jl:369
cov(x::AbstractArray{T,1} where T, y::AbstractArray{T,1} where T) in Base at statistics.jl:372
cov(X::Union{AbstractArray{T,1}, AbstractArray{T,2}} where T, Y::Union{AbstractArray{T,1}, AbstractArray{T,2}} where T, vardim::Int64) in Base at statistics.jl:382
cov(X::Union{AbstractArray{T,1}, AbstractArray{T,2}} where T, Y::Union{AbstractArray{T,1}, AbstractArray{T,2}} where T, vardim::Int64, corrected::Bool) in Base at statistics.jl:382
cov(x::AbstractArray{T,1} where T, Y::AbstractArray{T,2} where T) in Base at statistics.jl:385
cov(X::AbstractArray{T,2} where T, y::AbstractArray{T,1} where T) in Base at statistics.jl:386
cov(X::AbstractArray{T,2} where T, Y::AbstractArray{T,2} where T) in Base at statistics.jl:387

nalimilan replied:
You can just have `corrected=nothing` to detect whether a keyword argument has been left to its default. Then you just need to replace its value with `false` if it's equal to `nothing` after printing the warning using `Base.depwarn`.

nalimilan commented on src/moments.jl:
```diff
@@ -58,195 +68,231 @@ function var!(R::AbstractArray, A::RealArray, wv::WeightVec, dim::Int; mean=noth
+Return the `k`th order central moment of a real-valued array `v`, optionally
+specifying a weighting vector `wv` and a center `m`.
+"""
+moment(v::RealArray, k::Int, args...; kwargs...) = moment(v, Val{k}, args...; kwargs...)
```
That `Val` change is completely unrelated to this PR, please leave it out. I also doubt it will increase performance (rather the contrary).

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+"""
+    @weights name
+
+generates a new generic weight type with specified `name`, which subtypes `AbstractWeights`
```
Initial uppercase, also use imperative "Generate". See guidelines at https://docs.julialang.org/en/stable/manual/documentation/.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+
+Computes a correction factor for calculating `var`, `std` and `cov` with `n` observations.
+If `corrected=true` this will return ``\\frac{1}{n - 1}``
+(ie: [Bessel's correction](https://en.wikipedia.org/wiki/Bessel's_correction)),
```
"i.e.", no need for colon.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+"""
+    cfactor(n::Integer, corrected=false)
+
+Computes a correction factor for calculating `var`, `std` and `cov` with `n` observations.
```
"Compute". Maybe say "bias correction". Same for weighted version.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+    cfactor(n::Integer, corrected=false)
+
+Computes a correction factor for calculating `var`, `std` and `cov` with `n` observations.
+If `corrected=true` this will return ``\\frac{1}{n - 1}``
```
"this will return" -> "return". Same below.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+Computes a correction factor for calculating `var`, `std` and `cov` with `n` observations.
+If `corrected=true` this will return ``\\frac{1}{n - 1}``
+(ie: [Bessel's correction](https://en.wikipedia.org/wiki/Bessel's_correction)),
+otherwise it will return ``\\frac{1}{n}``.
```
"(i.e. no correction)".

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+"""
+    AnalyticWeights(vs, wsum=sum(vs))
+
+Construct an `AnalyticWeights` vector with weight values `vs` and sum of weights `wsum`.
```
I think this was in the previous docstring, but it's a good time to improve it: "and sum of weights `wsum`" is weird. Would be clearer to say "A precomputed sum may be provided as `wsum`." (like for `std`). Same for other weight types.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+
+Construct an `AnalyticWeights` vector with weight values `vs` and sum of weights `wsum`.
+
+Analytic weights describe a non-random relative importance (usually between 0 and 1)
```
I'm not sure what "non-random" means. Why not say that weights reflect the precision of an observation and that they are proportional to the inverse of the observation's variance, which typically arises when a value is the average of several observations?

rofinn replied:
> proportional to the inverse of the observation's variance, which typically arises when a value is the average of several observations

I don't know if that's always true though. For example, I think analytic or reliability weights can also be applied for arbitrary weight schemas (e.g., exponential weighting discussed below). The reason for the "non-random" is that these weights assume that there may be some external structure to the weighting (unlike frequency weights); this structure could be the exponential weighting or the inverse variance of each mean observation.

rofinn replied:
Also, from the wiki page

> If the weights are instead non-random (reliability weights) ...

nalimilan replied:
I think we'd better keep very strict definitions for the three types we provide here, and have a non-corrected generic type like `Weights` for less clearly identified kinds of weights (like what Stata calls "importance weights").

I still don't understand what "non-random" means here, but maybe that's just me. I'm not opposed to keeping it as long as we make the description more precise.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+"""
+    aweights(vs)
+
+Construct an `AnalyticWeights` vector from a given array.
```
"from array `vs`"

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+``\\frac{1}{\sum w - \sum w / \sum {w^2}}``
+"""
+function cfactor(w::AnalyticWeights, ::Type{Val{true}})
+    s = sum(w)
```
`s = w.sum`

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+"""
+function cfactor(w::AnalyticWeights, ::Type{Val{true}})
+    s = sum(w)
+    sum_sn = 0.0
```
To fully ensure type stability, should be something like `zero((zero(eltype(w)) / one(s)) ^ 2)`.

rofinn replied:
That would throw an exception, but yeah something like `zero(eltype(w)) / one(typeof(s)) ^ 2` would be fine.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+function cfactor(w::AnalyticWeights, ::Type{Val{true}})
+    s = sum(w)
+    sum_sn = 0.0
+    for x in w
```
`@inbounds` would be useful.

rofinn replied:
Yeah, I thought the iteration protocol would do that for me, but then I found this https://github.com/JuliaLang/julia/issues/11350.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+        sum_sn += (x / s) ^ 2
+    end
+
+    1 / (s * (1 - sum_sn))
```
This doesn't match the formula given in the docstring. Shouldn't this simply be `1 / (s - s/vecnorm(w))`?

rofinn replied:
The accidentally wrote the division backwards in the formula. According to the wiki page it should be `1 / (V1 - (V2 / V1^2)` where `V1` is the sum of the weights and `V2` is the sum of the squared weights. If the weights are normalized to sum to 1 then we can simplify it to `1 / (1 - V2)`. The we just scale the denominator by the actual scale of the weights, so that we don't need the weights to be normalized.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+
+Construct a `FrequencyWeights` vector with weight values `vs` and sum of weights `wsum`.
+
+Frequency weights describe the number of cases (or frequency) in which each observation
```
"number of times" (and then remove "in which")?

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+``\\frac{1}{\sum{w} - 1}``
 """
-    weights(vs)
+cfactor(w::FrequencyWeights, ::Type{Val{true}}) = 1 / (sum(w) - 1)
```
`sum(w)` -> `w.sum`

rofinn replied:
Is that just to avoid the function call overhead?

ararslan replied:
Avoids recomputing the sum

rofinn replied:
But we have a method `sum(wv::AbstractWeights) = wv.sum`?

ararslan replied:
Oh, idk then

nalimilan replied:
Ah, OK. But `w.sum` is simpler and more explicit.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+
+Probability weights represent the inverse of the sampling probability for each observation,
+providing a correction mechanism for under- or over-sampling certain population groups.
+These weight may also be referred to as sampling weights.
```
Plural "weights".

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+``\\frac{n}{(n - 1) \sum w}`` where `n = length(w)`
+"""
+function cfactor(w::ProbabilityWeights, ::Type{Val{true}})
+    s = sum(w)
```
`s = w.sum`

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+"""
+    eweights(n, [λ])
+
+Constructs an `AnalyticWeights` vector with a desired length `n` and smoothing factor `λ`,
```
"Construct". "with length" is enough.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+    eweights(n, [λ])
+
+Constructs an `AnalyticWeights` vector with a desired length `n` and smoothing factor `λ`,
+where each element is set to ``λ * (1 - λ)^(1 - i)``.
```
`"element in position ``i`` "`

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+Constructs an `AnalyticWeights` vector with a desired length `n` and smoothing factor `λ`,
+where each element is set to ``λ * (1 - λ)^(1 - i)``.
 
+# Arguments
```
As noted in the guidelines, the "arguments" section isn't recommended, except when there are many arguments. Here, the description of `n` is redundant with what is said above. You can just keep the last sentence about λ in the main description or in a separate paragraph.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
-Base.getindex(wv::WeightVec, i) = getindex(wv.values, i)
-Base.size(wv::WeightVec) = size(wv.values)
+"""
+    eweights(n, [λ])
```
`λ=0.99`, and drop the brackets. Or maybe drop the default value, as it seems quite arbitrary and unlikely to be exactly what people need?

rofinn replied:
Yeah, I guess the 0.99 is pretty arbitrary.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,189 @@
+    n > 0 || throw(ArgumentError("cannot construct weights of length < 1"))
+    0 <= λ <= 1 || throw(ArgumentError("smoothing factor must be between 0 and 1"))
+    w0 = map(i -> λ * (1 - λ)^(1 - i), 1:n)
+    aweights(w0)
```
Are these really analytical weights? The docs should mention this if that's the case. Can you explain what exponential weights are used for?

rofinn replied:
I often use exponential weights to describe the relative importance of observation in temporal data. For example, I often want to get summary statistics about some recent data, but putting greater value on more recent observations (as they better reflect the "current" state of the system). Since, these types of weights definitely aren't probability or frequency weights I got the impression that analytic or reliability weights best fit this.  I'm fine with removing this a niche use case, but my understanding is that exponential weights are pretty common when working with temporal data.

nalimilan replied:
Hmm, these aren't strictly exponential weights. Maybe this means we should keep a generic `Weights` type... Feel free to reintroduce it (to replace `WeightsVec`), though I'm not sure what's the best name for them.

nalimilan commented on test/moments.jl:
```diff
@@ -104,19 +104,50 @@ wv = weights(ones(5) * 2.0)
+# AnalyticWeights
+@test var(x, aweights(ones(10)); corrected=true) ≈ var(x)
+
+w = aweights(rand(10))
```
Rather than applying the formulas, this should use fixed values and hardcode the result. Then I can check that it's correct in other software if you like.

rofinn replied:
>Then I can check that it's correct in other software if you like

That's a great idea, I'd really appreciate that! What kind of software were you thinking of testing against? This PR is probably big enough as it is, but I was wondering if it would make sense to automate testing against R using `RCall`.

nalimilan replied:
I was thinking of the survey and Hmisc R packages, of Stata and maybe SAS. I wouldn't worry about using RCall, we can just hardcode the results for a few cases.

nalimilan commented:
@JeffreySarnoff `AbstractWeight` should probably implement `setindex!` so that they can be modified if needed. @rofinn just added support for `getindex`, you could open another PR for `setindex!`.

rofinn commented on src/moments.jl:
```diff
@@ -30,23 +31,32 @@ whereas it's `length(x)-1` in `Base.var`. The impact is that this is not a
     if mean == 0
-        Base.varm!(R, A, wv,
-            Base.reducedim_initarray(A, dim, 0, eltype(R)), dim)
+        Base.varm!(
```
Oh, yeah, forgot to fix this one.

rofinn commented on src/moments.jl:
```diff
@@ -58,195 +68,231 @@ function var!(R::AbstractArray, A::RealArray, wv::WeightVec, dim::Int; mean=noth
 
 ##### General central moment
+"""
+    moment(v, k, [wv::AbstractWeights], m=mean(v); corrected=false)
```
I was kind of wondering if it would make sense to just throw a warning when calling moment with k > 2. That way it serves as a reminder to come back to the issue. I also fine with removing it when I revert the `Type{Val{...}}` commit.

nalimilan replied:
Removing it is simpler, and it doesn't sound too useful to support and argument which only works in some cases. People can use `var` if they want the correction.

rofinn commented on src/StatsBase.jl:
```diff
@@ -17,14 +17,22 @@ module StatsBase
+    FrequencyWeights,   # the type for representing a frequency weight vectors
+    ProbabilityWeights, # the type for representing a probability/sampling weight vectors
+    ExponentialWeights, # the type for representing exponential weights
+    weights,            # alias for aweights
```
Don't we still want to export it (even if it's deprecated) to avoid breaking functionality?

nalimilan replied:
`@deprecate` automatically exports for this reason.

rofinn commented on src/cov.jl:
```diff
@@ -58,66 +58,41 @@ cov
 
 """
-    mean_and_cov(x, [wv::WeightVec]; vardim=1) -> (mean, cov)
+    mean_and_cov(x, [wv::AbstractWeights]; vardim=1) -> (mean, cov)
```
Yeah, I was holding off updating these docstrings until I was more comfortable with the behaviour. Since, folks don't seem to have an issue with the `corrected=false` keyword for all these functions I'll update that.

rofinn commented on src/hist.jl:
```diff
@@ -249,14 +247,15 @@ function append!{T,N}(h::AbstractHistogram{T,N}, vs::NTuple{N,AbstractVector})
     end
     h
 end
+
```
Yeah, that was a typo from fixing conflicts when I rebased with master.

ararslan commented:
I just want to say thanks so much for your hard work and perseverence here, @rofinn. You're doing fantastic work and I'm excited to see this get merged.

rofinn commented:
Is it possible to turn deprecation warnings on and off during testing in julia? It would be nice to check that I haven't broken any existing behaviour without producing a bunch of deprecation warning during testing.

nalimilan commented:
> Is it possible to turn deprecation warnings on and off during testing in julia? It would be nice to check that I haven't broken any existing behaviour without producing a bunch of deprecation warning during testing.

`julia --depwarn=no` is what you need.

nalimilan commented on src/cov.jl:
```diff
@@ -44,80 +37,73 @@ that the data are centered and hence there's no need to subtract the mean.
 When `vardim = 1`, the variables are considered columns with observations in rows;
 when `vardim = 2`, variables are in rows with observations in columns.
 """
-function scattermat end
```
This line was actually correct since the docstring documents several methods, not just the following one. Same below.

rofinn replied:
Right, but is that line even necessary (calling help on the method name still works the same). Also, the other line is just `cov` vs `function cov end`, what would the preference be to maintain consistency?

ararslan replied:
`cov` refers to an existing object; `function cov end` creates a function with 0 methods. Typically the latter is only used for forward-declaring functions to applying docstrings. I think it's usually preferable to use the former when the function has already been defined.

nalimilan replied:
In general it doesn't really make a big difference whether you attach the docstring to a method or to the function, but it matters in some cases: the docstring for the function will (or should) always appear first, and when a link to the source is provided (in the online manual) it wouldn't be really correct to point to a particular method.

rofinn replied:
Ah, okay, to clarify, the preference moving forward is that we want the docstring attached to `function <name> end` with all other methods below that?

rofinn replied:
> No, just keep the existing organization. Always better to avoid making changes unrelated to the PR, else it's very hard to review. For example, moving scattermat_zm makes it very hard to see where its code changed.

Fair point. I do tend to get carried away when I'm editing code.

nalimilan commented on src/cov.jl:
```diff
@@ -44,80 +37,73 @@ that the data are centered and hence there's no need to subtract the mean.
-        mean == nothing ? scattermat_zm(x .- Base.mean(x, wv, vardim), wv, vardim) :
-        scattermat_zm(x .- mean, wv, vardim)
-    end
+* AnalyticWeights: ``\\frac{1}{\sum w - \sum {w^2} / \sum{w}^2}``
```
Formula is incorrect AFAICT.

rofinn replied:
I guess it should be ``\\frac{1}{\sum w - \sum {w^2} / \sum w}``, which wouldn't be a problem if `w` was prenormalized, but yeah.
```julia
julia> cvar1(w) = sum(w) * (1 - sum(normalize(w, 1) .^ 2))
cvar1 (generic function with 1 method)

julia> cvar2(w) = sum(w) - (sum(w .^2) / sum(w))
cvar2 (generic function with 1 method)

julia> w = rand(10)
10-element Array{Float64,1}:
 0.608948
 0.980859
 0.656496
 0.994447
 0.912615
 0.654867
 0.0928164
 0.308836
 0.782151
 0.952136

julia> cvar1(w)
6.132433099237098

julia> cvar2(w)
6.132433099237098
```

rofinn commented:
Alright, apart from a couple remaining points about how to use @deprecate_binding and documenting groups of methods with `function <name> end` that's all the code changes...? I think I just need to update the documentation now if folks are mostly alright with this? @nalimilan I've updated the `tests/moments.jl` file to use `@testset` and hard coded values for the corrected variances, but I'd really appreciate if you could double check those values make sense.

nalimilan commented on test/weights.jl:
```diff
@@ -3,355 +3,376 @@ using Base.Test
+    @testset "eweights" begin
+        λ = 0.2
+        wv = eweights(4, λ)
+        @test round(values(wv), 4) == [0.2, 0.25, 0.3125, 0.3906]
```
Copy the full precision numbers rather than rounding. That could help spotting precision issues if somebody tweaks the formula.

Should also check the type of `wv`.

nalimilan commented on test/weights.jl:
```diff
@@ -3,355 +3,376 @@ using Base.Test
 
-## the sum and mean syntax
+    @testset "Sum" begin
+        @test sum([1.0, 2.0, 3.0], fweights([1.0, 0.5, 0.5])) ≈ 3.5
```
Can you put this inside a loop and test all weights types? Same below (and everywhere this applies).

rofinn replied:
I'll note that this causes tests to take significantly longer to run (particularly `test/weights.jl`). Would it make sense to reduce the median and quantile testsets slightly?

nalimilan replied:
Unless some tests are really redundant, I'd rather have too many tests than too few of them.

rofinn commented on src/weights.jl:
```diff
@@ -2,44 +2,191 @@
+"""
+    eweights(n, λ)
+
+Construct an `AnalyticWeights` vector with length `n`,
```
Wouldn't​ those weights be uncorrected though? I'd like the exponential weights to support correction.

nalimilan replied:
But what kind of correction? IIUC exponential weights do not correspond to a well-defined variance structure like analytic weights, so the correction used for them does not apply.

rofinn replied:
Hmm, I see what you mean. Would it make sense to go back to using frequency weights here, as they make the fewest assumption about the weighting structure?

nalimilan replied:
I don't think that would be correct either, as the assumptions are also very strong. The `n-1` correction doesn't apply for exponential weights either.

rofinn replied:
I could also just leave the `eweights` out of this PR and maybe we can come back to this conversation. AFAICT, stata has something called `iweights`, but I don't really know much about them.

nalimilan replied:
We could leave exponential weights for another PR, but we should decide whether to have a generic weights type or not now, before deprecating `weights` and `WeightsVec`.

`iweights` in Stata are just a custom kind of weights without any precise definition. This is indeed the kind of thing we seem to need, though the function would better be called `weights`.

rofinn replied:
I'll keep the `weights` definition around and just deprecate the `WeightVec` binding, but I'm going to remove the exponential weights code for now.

rofinn commented on src/weights.jl:
```diff
@@ -2,44 +2,191 @@
+    s = w.sum
+
+    if corrected
+        sum_sn = zero(eltype(w)) / one(typeof(s)) ^ 2  # to ensure type stability
```
Wow, I've never seen that RHS thing before.
```
julia> @code_native zero((zero(Float64) / one(Int)) ^ 2)
	.section	__TEXT,__text,regular,pure_instructions
Filename: number.jl
	pushq	%rbp
	movq	%rsp, %rbp
Source line: 61
	xorps	%xmm0, %xmm0
	popq	%rbp
	retq
Source line: 61
	nopl	(%rax)

julia> @code_native (zero(Float64) / one(Int)) ^ 2
	.section	__TEXT,__text,regular,pure_instructions
Filename: math.jl
	pushq	%rbp
	movq	%rsp, %rbp
Source line: 358
	movslq	%edi, %rax
	cmpq	%rdi, %rax
	jne	L25
	movabsq	$__powidf2, %rax
	popq	%rbp
	jmpq	*%rax
L25:
	movabsq	$jl_throw, %rax
	movabsq	$4650306176, %rdi       ## imm = 0x1152E0A80
	callq	*%rax
	nop
```

nalimilan replied:
Come to think of it, this can be written much more simply as `sum_sn = sum(x -> (x/s)^2, w)`. (BTW, what does "_sn" stand for?)

rofinn replied:
sn = squared norm

rofinn replied:
@nalimilan `sum_sn = sum(x -> (x/s)^2, w)` doesn't seem to be a performant.
```
julia> @time sum_sn = sum(x -> (x / s) ^ 2, w)
  0.031813 seconds (12.23 k allocations: 563.890 KB)
...
julia> @time begin
           sum_sn = zero((zero(eltype(w)) / one(typeof(s))) ^ 2)  # to ensure type stability
           @inbounds for x in w
                   sum_sn += (x / s) ^ 2
               end
       end
  0.000020 seconds (55 allocations: 1.016 KB)
```

nalimilan replied:
The second benchmark is wrong, since it doesn't return `sum_sn` the loop is optimized out. Wrapping everything in functions to make sure the closure is dealt with properly, I get this:
```julia
julia> function f1(w, s)
          sum_sn = zero((zero(eltype(w)) / one(typeof(s))) ^ 2)  # to ensure type stability
          @inbounds for x in w
              sum_sn += (x / s) ^ 2
          end
          sum_sn
       end
f1 (generic function with 1 method)

julia> f2(w, s) = sum(x -> (x / s) ^ 2, w)
f2 (generic function with 1 method)

julia> w = collect(1:100_000);

julia> @benchmark f1($w, $(sum(w)))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     868.741 μs (0.00% GC)
  median time:      870.310 μs (0.00% GC)
  mean time:        894.750 μs (0.00% GC)
  maximum time:     1.217 ms (0.00% GC)
  --------------
  samples:          5580
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

julia> @benchmark f2($w, $(sum(w)))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     97.496 μs (0.00% GC)
  median time:      99.200 μs (0.00% GC)
  mean time:        103.157 μs (0.00% GC)
  maximum time:     380.056 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%

```

I'm not sure why `sum` is faster in this case...

rofinn commented on src/cov.jl:
```diff
@@ -44,80 +37,73 @@ that the data are centered and hence there's no need to subtract the mean.
 """
-function mean_and_cov end
-
+    cov(X, wv::AbstractWeights, [vardim, corrected])
```
The problem with making `corrected` a keyword argument is that I can't deprecate the behaviour because I don't know if corrected has been set. Also, in base `cov` doesn't have `corrected` as a keyword argument, just `std` and `var`.

nalimilan replied:
As I said in other comments, use `; corrected::Union{Bool, Void}=nothing` and check `corrected === nothing` before printing the warning.

nalimilan replied:
For `cov`, it's OK to copy Base and use a positional argument.

rofinn commented on src/deprecates.jl:
```diff
@@ -43,3 +43,146 @@ findat(a::AbstractArray, b::AbstractArray) = findat!(Array{Int}(size(b)), a, b)
+
+Construct a `WeightVec` with weight values `vs` and sum of weights `wsum`.
+"""
+function WeightVec{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs))
```
I'm confused, this is in `deprecates.jl`.

nalimilan replied:
Sorry, I'm not sure how I missed that...

rofinn commented on src/deprecates.jl:
```diff
@@ -43,3 +43,146 @@ findat(a::AbstractArray, b::AbstractArray) = findat!(Array{Int}(size(b)), a, b)
+    string(fname, " will default to `corrected=true` in the future.")
+
+# The following methods are for wrapping the deprecated `correction=false` behaviour.
+# When we default to `correction=true` these methods should be removed in favour of
```
I figured it would be better to not use unions and have methods that can simply be deleted in the future. I considered the `Union{Bool, Void}` pattern, but that seemed like it didn't separate the deprecated behaviour as clearly.

nalimilan replied:
It's not pretty, but the priority must be the final public API, not the intermediate state nor the internal ugliness.

rofinn replied:
Okay.

rofinn commented on test/deprecates.jl:
```diff
@@ -0,0 +1,201 @@
+
+@testset "StatsBase.Deprecates" begin
+
+@testset "Deprecates WeightVec and weights" begin
```
It seems like a good idea to test deprecated features as it ensures the deprecations don't break functionality. Adding these tests actually helped me catch a few issues.

nalimilan replied:
Unfortunately, doing this will print lots of deprecation warnings, which will make the output hard to read, and deprecations from Base that need to be fixed won't be easy to spot. As long as you've tested it once locally, it's OK.

rofinn replied:
Would it make sense to keep the `ENV["TEST_DEPRECATES"]` behaviour around to make testing locally easier?

nalimilan replied:
I'd rather not. Deprecated code is supposed to be removed relatively soon anyway.

rofinn replied:
Okay, I'll add an item to my checklist to remove it before dropping the "[WIP]" from this PR, but please just ignore that until then.

rofinn commented on src/moments.jl:
```diff
@@ -160,39 +270,39 @@ function _moment3(v::RealArray, m::Real)
         s += z * z * z
     end
-    s / n
+    varcorrection(n, false) * s
```
That's why I've set it to `corrected=false` which is equivalent to `s / n` or `(s / sum(w))`. If you'd prefer I can keep the `varcorrection` usage to just `_moment2`.

nalimilan replied:
Yes, I'd prefer that. I don't see the point in calling a function which is known to be incorrect just to get `1/n`.

rofinn commented on src/moments.jl:
```diff
@@ -1,52 +1,100 @@
+
+Here we calculate the biased weighted variance (`corrected=false`) as:
+
+``\\frac{1}{\sum{w}} \sum\limits_{i = 1}^N {w_i\left( {x_i - \bar x} \right)^2 }``
```
Does anyone mind if I switch over to using "```math" since we're on 0.5 and higher.

rofinn replied:
According to the link on 0.5 and higher we don't need any extra backslashes?

rofinn replied:
Also, it looks like we already aren't using the extra backslash.

https://github.com/JuliaStats/StatsBase.jl/blob/master/src/scalarstats.jl#L45

rofinn replied:
Before I update all the appropriate methods how does this look?

```
"""
    varm(x, w::AbstractWeights, m, [dim]; corrected=false)

Return the variance of a real-valued array `x` with a known mean `m`, optionally
over a dimension `dim`. Observations in `x` are weighted using weight vector `w`.

The uncorrected (when corrected=false) sample variance is defined as:

``\\frac{1}{\\sum{w}} \\sum{i=1}^n {w_i\\left({x_i - m}\\right)^2 }``
where n = `length(w)`.

The unbiased estimate of the population variance is computed by replacing
``\\frac{1}{\\sum{w}}`` with a factor dependent on the type of weights used:

* AnalyticWeights: ``\\frac{1}{\\sum w - \\sum {w^2} / \\sum w}``
* FrequencyWeights: ``\\frac{1}{\\sum{w} - 1}``
* ProbabilityWeights: ``\\frac{n}{(n - 1) \\sum w}``
"""
```

ararslan replied:
Content wise I think that's good. A couple of purely cosmetic notes: I'd probably say "compute" instead of "return," and I'd wrap the longer formula in a displayed math block:

    ```math
    \\frac{1}{\\sum w} \\sum_{i=1}^n w_i \left( x_i - m \right)^2
    ```
    where ``n`` is the length of the input.

and I'd code format `corrected=false`.

rofinn replied:
Thanks.

nalimilan replied:
Looks good, but you need backticks around `corrected=false`.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,191 @@
+"""
+    varcorrection(w::ProbabilityWeights, corrected=false)
+
+``\\frac{n}{(n - 1) \sum w}`` where `n = length(w)`
```
While we're at it, we should use the more correct `n = count(!iszero, w)`. The code and the other docstring will need to be adjusted. To test this, you could simply add an element with a zero weight to the data: the computed variance must remain the same.

As a future optimization, we could store the number of non-zero weights at construction, when the sum is computed, but that's not a priority for this PR.

rofinn commented on test/cov.jl:
```diff
@@ -1,6 +1,7 @@
 using StatsBase
 using Base.Test
 
+@testset "StatsBase.Covariance" begin
```
I copied the naming convention from [test/hist.jl](https://github.com/JuliaStats/StatsBase.jl/blob/master/test/hist.jl), but I imagine once all the tests have been converted to using `@testset` this should be changed.

nalimilan replied:
OK, not a big deal. I have no idea what the convention is.

rofinn commented:
Alright, I think that addresses all the comments from the last review iteration. Please ignore the `test/deprecates.jl` and the corresponding changes in `runtests.jl` as I'm going to remove that once this PR is no longer a WIP.

nalimilan commented on src/common.jl:
```diff
@@ -26,3 +26,19 @@
+function depcheck(fname::Symbol, b::DepBool; msg::AbstractString=CORRECTED_DEP_MSG,
+                  default::Bool=false)
+    if b == nothing
+        Base.depwarn(string(fname, " ", msg), fname)
```
If you define the message here, you can just do `"$fname will default..."`.

nalimilan commented on src/common.jl:
```diff
@@ -26,3 +26,19 @@
+                                 "Use corrected=false for previous behaviour.")
+
+function depcheck(fname::Symbol, b::DepBool; msg::AbstractString=CORRECTED_DEP_MSG,
+                  default::Bool=false)
```
`default` is never used AFAICT, so better drop it.

nalimilan commented on src/cov.jl:
```diff
@@ -48,76 +49,60 @@ function scattermat end
+    scattermat_zm(x .- mean, vardim)
 
-    ## weighted cov
-    Base.cov(x::DenseMatrix, wv::WeightVec; mean=nothing, vardim::Int=1) =
```
Do not remove this method. Also, the docstring needs to mention `mean=nothing` again.

rofinn replied:
That was only supported being used when VERSION < v"0.5.0-dev+679" which we aren't supporting anymore anyways.

nalimilan commented on src/cov.jl:
```diff
@@ -48,76 +49,60 @@ function scattermat end
+matrix (`corrected=false`) can be computed by multiplying `scattermat(X, wv)` by
+``\\frac{1}{\\sum{w}}`` to normalize. However, the unbiased covariance matrix
+(`corrected=true`) is dependent on the type of weights used:
+
```
Does Markdown require an empty line before a bullet list? Else, remove it.

nalimilan commented on src/cov.jl:
```diff
@@ -48,76 +49,60 @@ function scattermat end
-is equivalent to `scattermat(X, wv) / sum(wv)`.
+Compute the weighted covariance matrix. Similar to `var` and `std` the biased covariance
+matrix (`corrected=false`) can be computed by multiplying `scattermat(X, wv)` by
+``\\frac{1}{\\sum{w}}`` to normalize. However, the unbiased covariance matrix
```
`w` has not been defined. Maybe replace `wv` with `w` everywhere it makes sense?

nalimilan commented on src/cov.jl:
```diff
@@ -48,76 +49,60 @@ function scattermat end
-    end
+function mean_and_cov(x::DenseMatrix, vardim::Int=1; corrected::Bool=true)
+    m = mean(x, vardim)
+    return m, Base.covm(x, m, vardim, corrected)
```
Shouldn't this be `corrected=corrected`? If so, that means that function wasn't tested at all?

rofinn replied:
This is calling the method in Base which uses a positional argument.

nalimilan commented on src/moments.jl:
```diff
@@ -1,52 +1,78 @@
+over a dimension `dim`. Observations in `x` are weighted using weight vector `w`.
+
+The uncorrected (when `corrected=false`) sample variance is defined as:
 
```
Remove empty line (same in other places).

nalimilan commented on src/moments.jl:
````diff
@@ -1,52 +1,78 @@
-over a dimension `dim`. The weighting vector `wv` specifies frequency weights
-(also called case weights) for the result.
+```math
+\\frac{1}{\\sum{w}} \\sum{i=1}^n {w_i\\left({x_i - m}\\right)^2 }
````
`sum_{i=1}`

nalimilan commented on src/moments.jl:
```diff
@@ -1,52 +1,78 @@
+The unbiased estimate of the population variance is computed by replacing
+``\\frac{1}{\\sum{w}}`` with a factor dependent on the type of weights used:
+
+* AnalyticWeights: ``\\frac{1}{\\sum w - \\sum {w^2} / \\sum w}``
```
Missing backticks around type names (same in other docstrings).

nalimilan commented on src/moments.jl:
```diff
@@ -1,52 +1,78 @@
+    var(x, wv::AbstractWeights, [dim]; mean=nothing, corrected=false)
+
+Compute the variance of a real-valued array `x`, optionally over a dimension `dim`.
+Observations in `x` are weighted using weight vector  `wv`.
```
Double space to remove.

nalimilan commented on src/moments.jl:
```diff
@@ -1,52 +1,78 @@
-whereas it's `length(x)-1` in `Base.var`. The impact is that this is not a
-weighted estimate of the population variance based on the sample; it's the weighted
-variance of the sample.
+The unbiased estimate of the population variance is computed by replacing
```
``"(when `corrected=true`)"``

nalimilan commented on src/moments.jl:
```diff
@@ -58,99 +84,184 @@ function var!(R::AbstractArray, A::RealArray, wv::WeightVec, dim::Int; mean=noth
+    std(v, wv::AbstractWeights, [dim]; mean=nothing, corrected=false)
+
+Compute the standard deviation of a real-valued array `x`,
+optionally over a dimension `dim`. Observations in `x` are weighted using weight vector `w`.
```
Choose between `w` and `wv`.

nalimilan commented on src/moments.jl:
```diff
@@ -58,99 +84,184 @@ function var!(R::AbstractArray, A::RealArray, wv::WeightVec, dim::Int; mean=noth
-`dim`, as a tuple. A weighting vector `wv` can be specified to weight the estimates.
-The weights are assumed to be frequency weights, also called case weights.
+`dim`, as a tuple. Observations in `x` can be weighted via `wv`. Finally, bias correction
+can be applied to the variance calculation if `corrected=true`.
```
"can be" -> "is/will be" (as noted in a previous comment -- there are probably several occurrences of this)

nalimilan commented on src/moments.jl:
```diff
@@ -58,99 +84,184 @@ function var!(R::AbstractArray, A::RealArray, wv::WeightVec, dim::Int; mean=noth
-function _moment2(v::RealArray, m::Real)
+
+##### General central moment
+function _moment2(v::RealArray, m::Real; corrected=false)
```
Drop `corrected` from here, anyway it's undocumented and not used by `moment`. Same below.

rofinn replied:
This is called by `varm(x, w::AbstractWeights, m, [dim]; corrected=false)` earlier in the file.

nalimilan commented on src/moments.jl:
```diff
@@ -174,7 +285,7 @@ function _moment3(v::RealArray, wv::WeightVec, m::Real)
 end
 
-function _moment4(v::RealArray, m::Real)
+function _moment4(v::RealArray, m::Real; corrected=false)
```
Same here (and unused).

nalimilan commented on src/moments.jl:
```diff
@@ -230,23 +341,25 @@ function moment(v::RealArray, k::Int, m::Real)
 
-moment(v::RealArray, k::Int) = moment(v, k, mean(v))
-moment(v::RealArray, k::Int, wv::WeightVec) = moment(v, k, wv, mean(v, wv))
+moment(v::RealArray, k::Int; corrected=true) = moment(v, k, mean(v))
```
Same here, you can keep the original versions.

nalimilan commented on src/sampling.jl:
```diff
@@ -641,7 +641,7 @@ whether sampling is performed with replacement and `order` dictates whether an
 """
 wsample!(a::AbstractArray, w::RealVector, x::AbstractArray; replace::Bool=true, ordered::Bool=false) =
-    sample!(a, weights(w), x; replace=replace, ordered=ordered)
+    sample!(a, fweights(w), x; replace=replace, ordered=ordered)
```
Since we decided to keep `weights`, no need to make a change here nor below.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,202 @@
+A precomputed sum may be provided as `wsum`.
+
+The `Weights` type describes a generic weights vector which does not support
+bias correction.
```
This isn't really the place to mention bias correction (for variance), there are other major cases where this makes a difference (e.g. regression models). Use a broader formulation like ``"which does not support all operations possible for [`FrequencyWeights`](@ref), [`AnalyticWeights`](@ref) and [`ProbabilityWeights`](@ref)"``.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,202 @@
+"""
+    varcorrection(w::Weights, corrected=false)
+
+Returns ``\\frac{1}{\sum w}`` when corrected is false and throws an `ArgumentError`
```
``"when `corrected=false`"``
``"if `corrected=true`"``

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,202 @@
+when corrected is true.
+"""
+function varcorrection(w::Weights, corrected::Bool=false)
+    corrected && throw(ArgumentError("Weights does not support bias correction."))
```
"Weights" could be ambiguous. Maybe "the Weights type does not support bias correction: use FrequencyWeights, AnalyticWeights or ProbabilityWeights if applicable".

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,202 @@
+Returns ``\\frac{1}{\sum w}`` when corrected is false and throws an `ArgumentError`
+when corrected is true.
+"""
+function varcorrection(w::Weights, corrected::Bool=false)
```
Add `@inline`.

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,202 @@
+Construct an `AnalyticWeights` vector with weight values `vs`.
+A precomputed sum may be provided as `wsum`.
 
+Analytic weights describe a non-random relative importance (usually between 0 and 1)
```
Should mention the relationship with variance and the fact that this generally corresponds to average (cf. my previous comment on this).

nalimilan commented on src/weights.jl:
```diff
@@ -2,44 +2,202 @@
+"""
+    varcorrection(w::AnalyticWeights, corrected=false)
+
+``\\frac{1}{\sum w - \sum {w^2} / \sum w}``
```
This only applies when `corrected=true`: need to give the formula when `corrected=false`. Same for other types.

nalimilan commented on src/weights.jl:
```diff
@@ -345,10 +503,10 @@ end
 
 Compute the weighted median of an array `v` with weights `w`, given as either a
-vector or `WeightVec`.
+vector or an `AbstractWeights` object/vector.
```
Sorry, I meant either "object" or "vector", not both. :-)

(Same below.)

nalimilan commented on test/moments.jl:
```diff
@@ -1,122 +1,284 @@
+
+@testset "Variance and Standard Deviation" begin
+    @testset "Vectors" begin
+        x = [0.57, 0.10, 0.91, 0.72, 0.46]
```
Can you add an observation with a `0` weight to test their handling? Shouldn't change the result.

nalimilan commented on test/weights.jl:
```diff
@@ -3,355 +3,370 @@ using Base.Test
-
-## wsum along dimensions
+@testset "StatsBase.Weights" begin
+weight_funcs = (aweights, fweights, pweights)
```
`weights` is missing.

nalimilan commented on test/weights.jl:
```diff
@@ -3,355 +3,370 @@ using Base.Test
+@testset "StatsBase.Weights" begin
+weight_funcs = (aweights, fweights, pweights)
+
+@testset "Construction" begin
```
Could you move as many of the port to `@testset` to a separate PR which wouldn't change anything else? Shouldn't be hard at least for this file, and it would make this PR cleaner.

rofinn replied:
Well we're already looping over weight types fairly often, so I don't think it'll help that much. Also, it seems pretty awkward to only have parts of files using `@testet` and will make the test output more verbose.

nalimilan replied:
My point was to make it easier to see what actually changed in the commits vs. what is just a reorganization wrapped inside `@testset`. This is important for review but also for being able to track changes in git history later. Then adding a loop over each weight type will mostly introduce spacing changes, which are visible as such in the diff.

nalimilan replied:
Can you open that PR? That would really make this one cleaner before we merge it.

rofinn replied:
@nalimilan Is it alright if I keep the `@testset`s when they're logically tied to my new tests or the ones wrapping the entire file (they aren't required, but should change the diff much and will keep the test output clean)?

nalimilan replied:
It's OK to keep them when you modified tests so much that it would be painful to separate the changes, but please move as much changes as possible to a new PR. Even when changes are limited, it's not always easy to distinguish mechanical changes from more subtle ones (`diff` and GitHub are easily confused).

nalimilan commented on docs/source/weightvec.rst:
```diff
@@ -47,16 +47,37 @@ Let ``w`` be an instance of ``WeightVec``:
 - Statistical functions that utilize weights often need the sum of weights for various purposes. The weight vector maintains the sum of weights, so that it needn't be computed repeatedly each time the sum of weights is needed.
 
+
+Other AbstractWeights types and bias correction
```
"other" doesn't make a lot of sense here since no other type has been introduced. It would be nice to have a section giving a short summary of the four concrete types of weights (mostly copy/paste of the docstrings). Then this section could just be called "Variance bias correction", and it could also mention that no correction is possible for `Weights` because it isn't specific enough.

BTW, needs backticks around `AbstractWeights` even in headings IIRC.

nalimilan commented on src/weights.jl:
```diff
@@ -57,7 +57,8 @@ Construct a `Weights` vector with weight values `vs`.
 The `Weights` type describes a generic weights vector which does not support
-bias correction.
+all operations possible for [`FrequencyWeights`](@ref), [`AnalyticWeights`](@ref)
+and [`ProbabilityWeights`](@ref)"
```
" -> .

rofinn commented:
I've rebased with master and updated the docs again. I opted for making each weight type a subheading under "Variance bias correction" in the docs. If @nalimilan and @ararslan think this is just about read to merge I'll drop my deprecation tests and remove the "[WIP]".

rofinn commented on docs/source/sampling.rst:
```diff
@@ -10,9 +10,9 @@ The package provides functions for sampling from a given population (with or wit
+.. function:: sample(a, n[; replace=true, ordered=false])
 
-    Randomly draw ``n`` elements from ``a``. 
+    Randomly draw ``n`` elements from ``a``.
```
Sorry, I think these are from my editor automatically removing extra whitespace.

ararslan replied:
That's fine, a lot of editors do that. Makes the diff marginally harder to review, but I think it's worth having cleaner whitespace.

rofinn commented on docs/source/weightvec.rst:
```diff
@@ -47,16 +47,84 @@ Let ``w`` be an instance of ``WeightVec``:
 
-Why we want a WeightVec type
------------------------------
+Why we want an AbstractWeights type
```
This didn't initially have backticks, but should I add them

nalimilan replied:
As you like. You can't fix all typos in the code, but this line is quite close to the ones you touched, so...

nalimilan commented on docs/source/weightvec.rst:
```diff
@@ -74,10 +74,57 @@ vs
-- ``AnalyticWeights``: :math:`\frac{1}{\sum w - \sum {w^2} / \sum w}`
-- ``FrequencyWeights``: :math:`\frac{1}{\sum{w} - 1}`
-- ``ProbabilityWeights``: :math:`\frac{n}{(n - 1) \sum w}` where ``n`` equals `count(!iszero, w)`
+Weights
```
It's weird to introduce the different kinds of weights inside the section on variance. I'd rather define them first in a separate section, and then explain what difference this makes for the correction of variance. I'm not even sure you need to copy all formulas: referring to the docstrings for `var` can be enough.

nalimilan replied:
There's also a section above which introduces generic weights without mentioning the other types of weights. It would be good to adapt it, since typically people should use a specific type rather than the generic one.

nalimilan commented on docs/source/weightvec.rst:
```diff
@@ -63,6 +50,19 @@ Probability weights represent the inverse of the sampling probability for each o
+.. code-block:: julia
+
+    w = Weights([1., 2., 3.])
+    w = Weights([1., 2., 3.,], 6.)
```
If you keep the two-argument version for this constructor, you should show it for all types. Or just remove it here.

rofinn replied:
I don't see the 2 argument version being used often, so I'll just remove it.

nalimilan commented:
Any comments about [this](https://github.com/JuliaStats/StatsBase.jl/pull/250#discussion_r114652368) and [this](https://github.com/JuliaStats/StatsBase.jl/pull/250#commitcomment-22021458)?

rofinn commented:
Sorry, I missed your response to the first one and was trying to think of a solution for the second.

nalimilan commented on src/moments.jl:
```diff
@@ -1,52 +1,70 @@
+``\\frac{1}{\\sum{w}}`` with a factor dependent on the type of weights used:
+* `AnalyticWeights`: ``\\frac{1}{\\sum w - \\sum {w^2} / \\sum w}``
+* `FrequencyWeights`: ``\\frac{1}{\\sum{w} - 1}``
+* `ProbabilityWeights`: ``\\frac{n}{(n - 1) \\sum w}`` where ``n`` equals `count(!iszero, w)`
```
Ah, and it would be cool to mention that no correction is possible for generic `Weights`.

nalimilan commented:
OK, let's go with this. Can you remove the tests for deprecated features and the hacks that go with it?

nalimilan commented on docs/source/cov.rst:
```diff
@@ -22,14 +22,14 @@ This package implements functions for computing scatter matrix, as well as weigh
+    Weighted scatter matrix. The weights are given by a weight vector ``wv`` of type ``AbstractWeights`` (see :ref:`weightvec`).
 
-.. function:: cov(X, wv[; vardim=..., mean=...])
+.. function:: cov(X, wv[; vardim=..., mean=..., corrected=...])
```
`wv` needs to be changed to `w` here too, and probably in a few other places.

nalimilan commented on test/runtests.jl:
```diff
@@ -17,6 +25,10 @@ tests = ["weights",
          "statmodels"]#,
          #"statquiz"]
 
+if !depwarns || test_deprecates
```
These are the "hacks" I referred to (there's another place too).

rofinn replied:
Sorry, forgot to save the file.

rofinn commented:
NOTE: travis failed pulling down julia-0.5 dmg for 1 build.

ararslan commented on docs/source/counts.rst:
```diff
@@ -8,7 +8,7 @@ Counting over an Integer Range
 .. function:: counts(x, a:b[, wv])
 
-    Count the number of times (or total weights if a weight vector ``wv`` is given) values in ``a:b`` appear in array ``x``. Here, the optional argument ``wv`` should be a weight vector of type ``WeightVec`` (see :ref:`weightvec`).
+    Count the number of times (or total weights if a weight vector ``wv`` is given) values in ``a:b`` appear in array ``x``. Here, the optional argument ``wv`` should be a weight vector of type ``AbstractWeights`` (see :ref:`weightvec`).
```
Is `weightvec` still the right ReST reference?

rofinn replied:
Yeah, I didn't rename the file. I assume that's what the `:ref:` is for (I really don't know restructured text though)?

rofinn replied:
Yeah, I can confirm that the link still works (it's just pointing to the weightvec.rst file).

ararslan replied:
Awesome, thanks for checking! I don't know how it works either. 🙃

nalimilan commented:
Sorry, it turns out you'll have to update the signatures and docstrings in the .rst files manually. We really need to move to the new Documenter.jl infrastructure.

rofinn commented:
I'm guessing I just need to update the docs for scalarstats.rst since there isn't API documentation in weightvec.rst? Also, should I still just copy/paste even though the documentation doesn't even match the old doc strings?

nalimilan commented:
AFAICT you need to update `scalarstats.rst` and `cov.rst`. Unfortunately you'll have to copy the docstrings and adapt the syntax manually since the RST syntax is slightly different from Markdown.

rofinn commented:
I guess I'm going to need to rebase again.

nalimilan commented:
Yeah, that's painful, sorry about that. Unfortunately, the new .rst docstrings need some more adjustments: some single backticks need to be changed to double (around Julia expressions), lists should use `*` rather than `-`, and the indent should be four spaces. At least this is what I understand looking at other files and at the RST syntax reference.

nalimilan commented:
Thanks!

If you're still willing to work on weighting, it would be great to use the new types in GLM.jl, which interprets weights as frequency weights and should therefore only accept `FrequencyWeights`. Support for other types of weights shouldn't be hard to add, at least for probability weights.

rofinn commented:
Awesome, thanks @nalimilan and @ararslan for putting so much work into reviewing this! Hopefully, my future PRs will be a bit smoother.

rofinn commented:
@nalimilan I'll take a look at GLM.jl, but I might not have time to get that PR ready till later this week.

tkelman commented on src/moments.jl:
```diff
@@ -58,99 +78,179 @@ function var!(R::AbstractArray, A::RealArray, wv::WeightVec, dim::Int; mean=noth
+Base.std(v::RealArray, w::AbstractWeights; mean=nothing, corrected::DepBool=nothing) =
+    sqrt.(var(v, w; mean=mean, corrected=depcheck(:std, corrected)))
+
+Base.stdm(v::RealArray, m::RealArray, dim::Int; corrected::DepBool=nothing) =
```
guess this signature was here before, but it's a bit of piracy, isn't it?

rofinn replied:
Yeah, that's been there for a while, but I didn't notice when I updated it. We had a discussion on #248 about it. I think the plan was to move this method to base julia and add a version check, but I haven't gotten around to doing it yet.

matthieugomez commented:
Great pull request! Thanks everyone for doing it. 
Just to be sure, are we sure about the plural form Weights (used in R) rather than the singular form Weight (used in SAS, Stata, Python)?

ararslan commented:
Yes, because a vector of weights contains more than one weight. 😉
